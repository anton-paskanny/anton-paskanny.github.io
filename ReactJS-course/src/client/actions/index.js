export const SEARCH_MOVIE = 'SEARCH_MOVIE'

export const searchMovie = (type, value) => ({
    type: SEARCH_MOVIE,
    payload: {
        type,
        value
    }
});