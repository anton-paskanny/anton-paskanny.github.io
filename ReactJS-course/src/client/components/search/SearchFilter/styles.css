.search-filter {
    position: relative;
    display: flex;
    justify-content: space-between;
    margin-top: 20px;
    z-index: 3;
}

.search-filter__desc {
    color: #fff;
}

.search-filter__inner-wrapper {
    display: flex;
    align-items: center;
    font-size: 12px;
}

.search-filter__filters {
    display: flex;
}

.search-filter__filter-btn {
    display: block;
    margin-left: 10px;
    padding: 5px 15px;
    border: none;
    background-color: var(--secondary-color);
}

.search-filter__filter-btn--active {
    background-color: var(--primary-color);
}

.search-filter__search-btn {
    padding: 5px 30px;
    font-size: 14px;
    border: none;
    background-color: var(--primary-color);
}

.search-filter__desc,
.search-filter__filter-btn,
.search-filter__search-btn {
    color: #fff;
    text-transform: uppercase;
}