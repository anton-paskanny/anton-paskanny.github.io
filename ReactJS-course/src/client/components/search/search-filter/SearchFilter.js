import React, { PureComponent } from 'react';


class SearchFilter extends PureComponent {
    render() {
        console.log("Render search filter");

        return (
            <div className="search-by">
                    <p>Search by</p>
                    <ul>
                        <li>
                            <button>Title</button>
                        </li>
                        <li>
                            <button>Genre</button>
                        </li>
                    </ul>
                    <button onSubmit={this.props.handleSubmit}>
                        Search
                    </button>
            </div>
        )
    }
}

export default SearchFilter;