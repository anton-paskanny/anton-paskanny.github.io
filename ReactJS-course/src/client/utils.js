export const URL_BASE = 'https://reactjs-cdp.herokuapp.com/movies';

export const SORT_BY_DEFAULT = 'release_date';
export const SORT_BY_CONFIG = [
  {
    name: SORT_BY_DEFAULT,
    fieldToSortBy: 'release_date',
  },
  {
    name: 'rating',
    fieldToSortBy: 'vote_average',
  },
];

export const SEARCH_BY_DEFAULT = 'title';
export const SEARCH_BY_CONFIG = [
  {
    name: SEARCH_BY_DEFAULT,
  },
  {
    name: 'genres',
  },
];

export const throttle = (callback, limit) => {
  let wait = false;
  return function () {
    if (!wait) {
      callback.call();
      wait = true;
      setTimeout(() => {
        wait = false;
      }, limit);
    }
  };
};
