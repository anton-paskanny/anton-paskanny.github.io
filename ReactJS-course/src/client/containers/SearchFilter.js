import { connect } from 'react-redux';
import { toggleSearch } from '../actions/searchFilter';

import SeachFilter from '../components/Search/SearchFilter/SearchFilter';

export const mapStateToProps = state => ({
    searchType: state.search.type
});

export const mapDispatchToProps = dispatch => ({
    toggleSearch(type) {
        dispatch(toggleSearch(type));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(SeachFilter);