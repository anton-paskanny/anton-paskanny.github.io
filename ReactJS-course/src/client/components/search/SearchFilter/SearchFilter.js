import React, { PureComponent } from 'react';

import styles from './styles.css';

class SearchFilter extends PureComponent {
    renderSearchByItems() {
        return this.props.searchByConfig.map((item) => {
            return (
                <li key={item.name} className='search-filter__filter-item'>
                    <button type="button" className={'search-filter__filter-btn' + (item.active ? ' search-filter__filter-btn--active': '')}>{item.name}</button>
                </li>
            )
        });
    }

    render() {
        return (
            <div className="search-filter">
                    <div className="search-filter__inner-wrapper">
                        <p className="search-filter__desc">Search by</p>
                        <ul className="search-filter__filters" onClick={this.props.handleSearchByChange}>
                            { this.renderSearchByItems() }
                        </ul>
                    </div>
                    <button className="search-filter__search-btn" type="submit">
                        Search
                    </button>
            </div>
        )
    }
}

export default SearchFilter;