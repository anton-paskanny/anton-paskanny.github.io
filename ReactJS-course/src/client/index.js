import React from 'react';
import ReactDOM from 'react-dom';

import App from './components/App';

import normalize from './normalize.css';
import styles from './styles.css';

ReactDOM.render(<App />, document.querySelector('#root'));