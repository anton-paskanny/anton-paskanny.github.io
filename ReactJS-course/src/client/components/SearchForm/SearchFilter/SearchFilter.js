// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';

import { SEARCH_BY_CONFIG } from '../../../utils';

type Props = {
  searchType: string,
  className: string,
  toggleSearch: (text: string) => void
}

const InnerWrapper = styled.div`
  display: flex;
  align-items: center;
  font-size: 12px;
`;

const SearchFiltersList = styled.ul`
  display: flex;
`;

const SearchDesc = styled.p`
  color: #fff;
  text-transform: uppercase;
`;

const SearchBtn = styled(SearchDesc)`
  padding: 5px 30px;
  font-size: 14px;
  border: none;
  background-color: var(--primary-color);
  -webkit-appearance: none
`;

const FilterItemBtn = styled(SearchDesc)`
  display: block;
  margin-left: 10px;
  padding: 5px 15px;
  border: none;
  background-color: var(${props => props.primary ? '--primary-color' : '--secondary-color'});
`;

class SearchFilter extends PureComponent<Props> {
    handleSearchByChange = (e: { target: HTMLLIElement, preventDefault: () => void }) => {
      e.preventDefault();

      if (e.target.classList.contains('search-filter__filter-btn--active')) {
        return;
      }

      this.props.toggleSearch(e.target.textContent);
    }

    renderSearchByItems() {
      return SEARCH_BY_CONFIG.map(item => (
                <li key={item.name} className='search-filter__filter-item'>
                  <FilterItemBtn primary={item.name === this.props.searchType ? true : false}>
                    {item.name}
                  </FilterItemBtn>
                </li>
      ));
    }

    render() {
      return (
          <div className={this.props.className}>
            <InnerWrapper>
              <SearchDesc>Search by</SearchDesc>
              <SearchFiltersList onClick={this.handleSearchByChange}>
                { this.renderSearchByItems() }
              </SearchFiltersList>
            </InnerWrapper>
            <SearchBtn type="submit">
              Search
            </SearchBtn>
          </div>
      );
    }
}

export default styled(SearchFilter)`
  position: relative;
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
  z-index: 3;
  color: #fff;
`;
