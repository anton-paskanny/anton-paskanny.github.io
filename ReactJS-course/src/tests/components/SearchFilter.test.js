import React from 'react';
import { shallow } from 'enzyme';

import SearchFilter from '../../client/components/SearchForm/SearchFilter/SearchFilter';

import { SEARCH_BY_DEFAULT } from '../../client/utils';

let component, props;

describe('Search Filter', () => {
    beforeEach(() => {
        props = {
            searchType: SEARCH_BY_DEFAULT,
            toggleSearch: jest.fn()
        }
        component = shallow(<SearchFilter {...props} />);
    });
    it('render component correctly', () => {
        expect(component).toMatchSnapshot();
    });
    it('should handle search change correctly', () => {
        const event = {
            target: {
                classList: {
                    contains: jest.fn().mockReturnValue(false)
                },
                textContent: 'genres'
            },
            preventDefault: jest.fn()
        }
        component.find('.search-filter__filters').simulate('click', event);
        expect(props.toggleSearch).toHaveBeenCalledWith(event.target.textContent);
    });

    it('should not handle search when the same active item is clicked', () => {
        const event = {
            target: {
                classList: {
                    contains: jest.fn().mockReturnValue(true)
                },
                textContent: SEARCH_BY_DEFAULT
            },
            preventDefault: jest.fn()
        }
        component.find('.search-filter__filters').simulate('click', event);
        expect(props.toggleSearch).not.toHaveBeenCalledWith(event.target.textContent);
    });
});



