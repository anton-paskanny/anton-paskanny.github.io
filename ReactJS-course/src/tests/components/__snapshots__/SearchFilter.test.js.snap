// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Filter render component correctly 1`] = `
<div
  className="search-filter"
>
  <div
    className="search-filter__inner-wrapper"
  >
    <p
      className="search-filter__desc"
    >
      Search by
    </p>
    <ul
      className="search-filter__filters"
      onClick={[Function]}
    >
      <li
        className="search-filter__filter-item"
        key="title"
      >
        <button
          className="search-filter__filter-btn search-filter__filter-btn--active"
          type="button"
        >
          title
        </button>
      </li>
      <li
        className="search-filter__filter-item"
        key="genres"
      >
        <button
          className="search-filter__filter-btn"
          type="button"
        >
          genres
        </button>
      </li>
    </ul>
  </div>
  <button
    className="search-filter__search-btn"
    type="submit"
  >
    Search
  </button>
</div>
`;
